### PR TITLE
feat(batchUploader): Batchuploader Client

### PR DESCRIPTION
**Pull request by JR-Morgan**
Added the BatchUploader SDK project to speckle sharp.

Right now, I've added it under a new solution folder `ConnectorCore`
(feedback welcome on the name)

image

Before Merge:
 - [x] ~~Can we make a feature release for these? including the nuget for the batchuploader SDK?~~ We can push `nugets/2.17.0-testBatchUploaderSDK` and `2.17.0-testBatchUploaderSDK` tags
 - [x] https://github.com/specklesystems/speckle-sharp/pull/2968
 - [x] Are we ok with the repo structure?
 - [ ] Any CI work?

After merge:
 - [x] Create wip/feature release of connector 
 - [ ] Once we are happy with the stability of the client, we can release the sdk as a nuget. We'll wait a while before releasing a nuget.

### CODE REVIEW

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs:208]: did this slip in by mistake?

Comment (JR-Morgan) [reply Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs]: No this was a change on `main` that wasn't merged to `dev` yet, now resolved by https://github.com/specklesystems/speckle-sharp/pull/2968

**Comment by JR-Morgan** [comment]
@AlanRynne Shall we include this PR as part of the 2.17 wip? or would you prefer I create a separate feature release?
I think I'd prefer to get it in to 2.17 wip.
But this pr should get review from all reviewers added before merge.

**Comment by AlanRynne** [comment]
Ok! feel free to push a tag onto your branch to get this done! remember you need to push a separate one for the nugets. Or just ask me and i'll do it 🚀

**Comment by JR-Morgan** [comment]
Lets try and get this reviewed and merged in before the next wip

File: ConnectorCSI/ConnectorCSIShared/CSIApplicationLifetimeController.cs
Comment (AlanRynne) [inline ConnectorCSI/ConnectorCSIShared/CSIApplicationLifetimeController.cs:1]: Did this file addition get included by mistake?

File: ConnectorRevit/ConnectorRevit2020/ConnectorRevit2020.csproj
Comment (AlanRynne) [inline ConnectorRevit/ConnectorRevit2020/ConnectorRevit2020.csproj:31]: I'm aware of why this is being referenced and used within the connector, and although not a real blocker, this is something I'm afraid we're going to regret later.

File: Directory.Build.props
Comment (AlanRynne) [inline Directory.Build.props:87]: Beautiful

**Review by AlanRynne — COMMENTED** [review]
Just a few comments/questions

File: ConnectorRevit/ConnectorRevit/Entry/App.cs
Comment (JR-Morgan) [reply ConnectorRevit/ConnectorRevit/Entry/App.cs:149]: As part of the contract of the constructor functions, we can document our exceptional cases.
As of right now, there are none I can see.

If we somehow fail to construct a `HTTPClient`, or get the current process id. Sounds like there are bigger problems than a bug with our batchuploader.

File: ConnectorRevit/ConnectorRevit2020/ConnectorRevit2020.csproj
Comment (JR-Morgan) [reply ConnectorRevit/ConnectorRevit2020/ConnectorRevit2020.csproj:31]: In what sense are we likely to regret this?
The connector is the entry point for our plugins, if we have a connector specific dependency, this is the place for it no?

File: ConnectorCSI/ConnectorCSIShared/CSIApplicationLifetimeController.cs
Comment (JR-Morgan) [reply ConnectorCSI/ConnectorCSIShared/CSIApplicationLifetimeController.cs:1]: Think so, this may have been experimental. @connorivy I'm guessing this should be removed from this PR?

File: ConnectorCore/BatchUploader.OperationDriver/BatchUploadOperationDriver.cs
Comment (AlanRynne) [reply ConnectorCore/BatchUploader.OperationDriver/BatchUploadOperationDriver.cs:66]: Please check the edge case where batchuploader could be running but no jobs have run.
Would revit just remain open?

File: ConnectorCSI/ConnectorCSIShared/CSIApplicationLifetimeController.cs
Comment (JR-Morgan) [reply ConnectorCSI/ConnectorCSIShared/CSIApplicationLifetimeController.cs:1]: Removed now

File: ConnectorCore/BatchUploader.OperationDriver/BatchUploadOperationDriver.cs
Comment (JR-Morgan) [reply ConnectorCore/BatchUploader.OperationDriver/BatchUploadOperationDriver.cs:66]: This is a known "error state" that we plan on handling more explicitly in the Batchuploader desktop app (since it manages the revit subprocess directly)

**Review by AlanRynne — APPROVED** [review]
All conversations resolved ♥️ Merging in!

**Pull request merged**
